File: lib/chibirigor/type.rb
# frozen_string_literal: true

module Chibirigor
  # Type carriers (data that represents a type). Part 1 has only these three.
  module Type
    # The type of "this exact value." e.g. Const[1], Const["hi"]
    Const = Data.define(:value) do
      def to_s = value.inspect
    end

    # A named class. e.g. Nominal[:Integer]
    Nominal = Data.define(:name) do
      def to_s = name.to_s
    end

    # "Unknown / unknowable" — the linchpin of gradual typing.
    Dynamic = Data.define do
      def to_s = 'untyped'
    end

    # When the type doesn't collapse to one. e.g. Integer | String
    Union = Data.define(:members) do
      def to_s = members.map(&:to_s).join(' | ')
    end

    # The shape of a hash: remembers a type per key. e.g. {foo: 1, bar: "a"}
    HashShape = Data.define(:fields) do
      def to_s = "{#{fields.map { |key, type| "#{key}: #{type}" }.join(', ')}}"
    end

    # An array remembered "type per position." e.g. [1, "a"]
    Tuple = Data.define(:elements) do
      def to_s = "[#{elements.map(&:to_s).join(', ')}]"
    end

    # A generic type with type arguments. e.g. Array[Integer] (an array whose
    # element type Elem is rounded to one). We forget the positions and just say
    # "the element is Elem" (the fallback from generics 5b/5c).
    Generic = Data.define(:name, :args) do
      def to_s = "#{name}[#{args.map(&:to_s).join(', ')}]"
    end

    module_function

    # Fold an array of types into one. Flatten nested Unions and drop duplicates.
    def union(types)
      flat = types.flat_map { |t| t.is_a?(Union) ? t.members : [t] }.uniq
      return Dynamic.new if flat.empty? # empty = dead end; fall back to the safe side
      return Dynamic.new if flat.any?(Dynamic) # any untyped makes the whole thing untyped (gradual)
      return flat.first if flat.size == 1

      Union[flat.freeze]
    end

    # Whether the type counts as an integer (true for both Const[1] and Nominal[:Integer]).
    def integerish?(type)
      (type.is_a?(Const) && type.value.is_a?(Integer)) || type == Nominal[:Integer]
    end
  end
end

File: impls/rust/src/type_.rs
/// The type carrier. Corresponds to `Chibirigor::Type` in the Ruby implementation.
#[derive(Debug, Clone, PartialEq)]
pub enum Type {
    /// A type representing a literal value itself. Example: Const(Value::Int(1))
    Const(Value),
    /// A named class. Example: Nominal("Integer")
    Nominal(String),
    /// The "unknown" type. The crux of gradual typing.
    Dynamic,
    /// A Union type. Example: Integer | String
    Union(Vec<Type>),
    /// The shape of a hash. Remembers the type per key. Example: {foo: 1, bar: "a"}
    HashShape(Vec<(String, Type)>),
    /// Remembers an array as a "type per position". Example: [1, "a"]
    Tuple(Vec<Type>),
}

/// The kinds of value a Const type can hold (corresponding to Ruby's main literals).
#[derive(Debug, Clone, PartialEq)]
pub enum Value {
    Int(i64),
    Float(f64),
    Str(String),
    Bool(bool),
    Symbol(String),
    Nil,
}

impl std::fmt::Display for Value {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Value::Int(n) => write!(f, "{n}"),
            Value::Float(n) => write!(f, "{n}"),
            Value::Str(s) => write!(f, "{s:?}"),
            Value::Bool(b) => write!(f, "{b}"),
            Value::Symbol(s) => write!(f, ":{s}"),
            Value::Nil => write!(f, "nil"),
        }
    }
}

impl std::fmt::Display for Type {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Type::Const(v) => write!(f, "{v}"),
            Type::Nominal(n) => write!(f, "{n}"),
            Type::Dynamic => write!(f, "untyped"),
            Type::Union(members) => {
                let parts: Vec<_> = members.iter().map(|t| t.to_string()).collect();
                write!(f, "{}", parts.join(" | "))
            }
            Type::HashShape(fields) => {
                let parts: Vec<_> = fields.iter().map(|(k, v)| format!("{k}: {v}")).collect();
                write!(f, "{{{}}}", parts.join(", "))
            }
            Type::Tuple(elements) => {
                let parts: Vec<_> = elements.iter().map(|t| t.to_string()).collect();
                write!(f, "[{}]", parts.join(", "))
            }
        }
    }
}

/// Fold multiple types into one. Flatten nested Unions and remove duplicates.
pub fn union(types: Vec<Type>) -> Type {
    let mut flat = Vec::new();
    for t in types {
        match t {
            Type::Union(members) => flat.extend(members),
            other => flat.push(other),
        }
    }
    flat.dedup();
    if flat.is_empty() {
        return Type::Dynamic;
    }
    if flat.iter().any(|t| t == &Type::Dynamic) {
        return Type::Dynamic;
    }
    if flat.len() == 1 {
        flat.remove(0)
    } else {
        Type::Union(flat)
    }
}

/// Reduce a type to a Ruby class name. Used to match keys in the dispatch table.
pub fn class_of(t: &Type) -> Option<&str> {
    match t {
        Type::Const(Value::Int(_)) => Some("Integer"),
        Type::Const(Value::Float(_)) => Some("Float"),
        Type::Const(Value::Str(_)) => Some("String"),
        Type::Const(Value::Bool(true)) => Some("TrueClass"),
        Type::Const(Value::Bool(false)) => Some("FalseClass"),
        Type::Const(Value::Symbol(_)) => Some("Symbol"),
        Type::Const(Value::Nil) => Some("NilClass"),
        Type::Nominal(n) => Some(n.as_str()),
        _ => None,
    }
}
